### PR TITLE
batches: emphasize distinction between PAT and global credentials in docs

### LINKED ISSUE #21715: Document user batch changes credentials aren't used for importing changesets
Started conversation in https://github.com/sourcegraph/customer/issues/378. In short: we don't clearly document that importing and syncing changesets requires a batch changes site credential token, and doesn't use user tokens.

Also make sure to add it to this page: https://docs.sourcegraph.com/batch_changes/how-tos/tracking_existing_changesets

See also: #21889; ideally we should address both together.

### LINKED ISSUE #21889: doc: unclear hierarchy between global and personal access token
> When a user doesn’t have any credentials stored, we try to use the global credentials, if there are any. 

is unclear in the docs

### PR DESCRIPTION
**Pull request by courier-new**
Closes #21715. Maybe closes #21889 too??

Documents that a global service account token is required to import and sync changesets and clarifies the hierarchy between the personal access token and the global credentials in a couple places.

I also moved the deprecated and now-removed fallback for admins for the Sourcegraph external services token to the bottom of the "Configuring credentials" page and updated the warning about not using it.

### CODE REVIEW

File: doc/batch_changes/explanations/permissions_in_batch_changes.md [hunk 1]
```diff
@@ -42,28 +42,17 @@ All users are automatically given read permissions to a batch change. Granular p
...
 - Closing a changeset
 
-For instructions on adding and managing your code host access tokens, please refer to "[Configuring credentials](../how-tos/configuring_credentials.md)".
+Some interactions currently require a [global service account token](../how-tos/configuring_credentials.md#global-service-account) to also be configured for the code host due to the way we resolve ownership of a changeset. These operations are:
 
-See these code host specific pages for which permissions and scopes the tokens require:
-
-- [GitHub](../../../admin/external_service/github.md#github-api-token-and-access) (requires `repo`, `read:org`, `user:email`, and `read:discussion`scope)
-- [GitLab](../../../admin/external_service/gitlab.md#access-token-scopes) (requires `api`, `read_repository`, and `write_repository` scope)
-- [Bitbucket Server](../../../admin/external_service/bitbucket_server.md#access-token-permissions) (requires write permissions on the project and repository level)
```
Comment (courier-new) [inline doc/batch_changes/explanations/permissions_in_batch_changes.md:58]: I removed this section because
1. it's redundant with "Configuring credentials", which also details the scope requirements for each code host and is already referenced from this section
2. it feels misleading to points users to the `admin/external_service` pages here now because those are about setting up the Sourcegraph-wide code host connection tokens, which explicitly don't work for Batch Changes anymore

File: doc/batch_changes/explanations/permissions_in_batch_changes.md [hunk 2]
```diff
@@ -42,28 +42,17 @@ All users are automatically given read permissions to a batch change. Granular p
 
 ## Code host interactions in Batch Changes
 
-Interactions with the code host are performed by Sourcegraph with your personal access token for that code host. These operations include:
+Interactions with the code host are performed by Sourcegraph with your [personal access token](../how-tos/configuring_credentials.md#for-yourself) for that code host, or a [global service account token](../how-tos/configuring_credentials.md#global-service-account), if no personal access token is configured. These operations include:
```
Comment (mrnugget) [inline doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: I like the distinction here and it already clarifies things, but I think we should maybe restructure this a bit and clearly state that a global service account token is **required** for batch changes to work properly. Without the token, changesets won't be synced.

In the current version it sounds like "Syncing a changeset with the codehost" is something optional, or a feature.

So I personally would restructure this section to pull your new paragraph ("Some interactions currently require") up and reword it to make it clear that this is base functionality.

File: doc/batch_changes/how-tos/tracking_existing_changesets.md [hunk 3]
```diff
@@ -4,6 +4,11 @@ Batch Changes allow you not only to [publish changesets](publishing_changesets.m
 
 <img src="https://sourcegraphstatic.com/docs/images/batch_changes/tracking_existing_changesets_overview.png" class="screenshot center">
 
+## Requirements
```
Comment (mrnugget) [inline doc/batch_changes/how-tos/tracking_existing_changesets.md:7]: Good call adding this here!

File: doc/batch_changes/explanations/permissions_in_batch_changes.md (hunk 2 above)
Comment (mrnugget) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: Another thought here: if we now hard-require a global service account then we should also add this to the [requirements docs page](https://docs.sourcegraph.com/batch_changes/references/requirements).

But reading through the docs here again I'm also wondering whether I'm getting something wrong: do we really require a global service account to sync changesets?

Don't we then have the problem that we require a global token for base functionality to work but then automatically allow users to fallback to that token?

That makes me think that there's a "hole" in our product: syncing changesets with personal access tokens is not possible. But in a setup where the admin doesn't want a global token (or not have users fall back to it) how would we sync changesets?

Maybe @eseliger can clear this up.

**Review by mrnugget — COMMENTED** [review]
In general approve the wording and restructuring, but I'm wondering about how our product works exactly :grimacing:

File: doc/batch_changes/explanations/permissions_in_batch_changes.md (hunk 2 above)
Comment (courier-new) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: I updates the requirements doc to mention configuring credentials. I can make it more aggressive if setting up that global service account token is really a hard requirement.

Comment (mrnugget) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: Even with webhooks configured the `Syncer` will run in the background and try to sync changesets. Less often, if there are regular webhook requersts coming in, but still running.

We can talk about this in the sync today, but maybe @eseliger can chime in before that so we're on the same page.

Comment (eseliger) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: > Another thought here: if we now hard-require a global service account then we should also add this to the requirements docs page.

Agreed

> But reading through the docs here again I'm also wondering whether I'm getting something wrong: do we really require a global service account to sync changesets?

We don't [_require_ it in the syncer](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@es/no-executor-queue/-/blob/enterprise/internal/batches/syncer/syncer.go?L489-490) and [in the executor when importing](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@es/no-executor-queue/-/blob/enterprise/internal/batches/reconciler/executor.go?L381-382), so no it is not a hard requirement. It is highly encouraged at the moment, but we could not remove it back then because it would've been the breaking change that users must define such a service account. Now they just unknowingly have a global service account (not much better, lol). Ultimately, it should become super simple and be either use the user token for all activities _or_ the site credential but never mix and match. We're not there yet, though. **But the true current state is you don't need it.**

> Even with webhooks configured the Syncer will run in the background and try to sync changesets. Less often, if there are regular webhook requersts coming in, but still running.

100% correct. We always operate under the assumption that webhooks aren't perfect and some events could be missed. So we still try to sync with the code host at least every 8h, if rate limits allow.

Comment (mrnugget) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: Thank you! So it sounds to me like we could probably reword this and say something like

"Sourcegraph uses a global service account to keep changesets up to date"
...
"If no service account is set, it will fall back to one of the tokens in a configured code host connection"
...
"We highly recommend to configure a global service account though, since the fallback will likely be deprecated in the future."

Comment (eseliger) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: SGTM 👍

Comment (courier-new) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: Wow what a whirlwind, thank you both for the explanations! I eagerly await the day we can devote to simplifying this. 🙂

I can roll with that wording. Just to clarify, though, if we can use the code host connection token as a fallback still, then I think I need to update this (pre-existing) section, too:

> ## Site admin fallback token
> 
> > WARNING: This fallback was deprecated in Sourcegraph 3.27 and removed as of Sourcegraph 3.29.
> 
> For site admins, when neither a personal access token nor global credentials are configured, Batch Changes will fall back to using the Sourcegraph-wide token for the [code host connection](../../admin/external_service.md). This makes it easier to try out Batch Changes, but be aware that this may result in changesets being created or changed with different permissions from your normal code host user.
> 
> Non-admin users are unable to apply batch changes without configuring access tokens.

It sounds like:

1. This fallback was *not* removed in 3.29 and will still work probably until we get that point of massively simplifying the credentials.
2. Non-admin users *are* able to apply batch changes without configuring access tokens, because *anyone* can use this as the fallback token.

Did I get that right?

Comment (eseliger) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: > This fallback was not removed in 3.29 and will still work probably until we get to that point of massively simplifying the credentials (though we can still warn that it will be deprecated).

It was removed - but this was always just for the scope of _writing_ to the code host, ie creating a PR, updating it, pushing code and such. Syncing/importing is still a special case and has not been deprecated. I think the messaging around that wasn't super clear.

> Non-admin users are able to apply batch changes without configuring access tokens, because anyone can use this as the fallback token.

They can not use the code host tokens for writing operations (create,..). They can only use the user token or global site credential token, if they exist. We will, however, fall back to code host tokens for syncing and importing, when there are no site credentials configured. 

So decision tree is the following:

Is the operation read-only (import and sync)?
- If yes: Use site credential, fall back to global code host tokens, otherwise bail
- If no (so it is any other, writing, reconciler operation): Use user credential, fall back to site credential, otherwise bail

Comment (courier-new) [reply doc/batch_changes/explanations/permissions_in_batch_changes.md:45]: > but this was always just for the scope of *writing* to the code host

Ahhh got it, yes, that was indeed the missing piece. The second part follows and makes sense then. Thanks Erik!!! 👍

**Comment by courier-new** [comment]
So I attempted to revamp the "Configuring credentials" page to really make it as obvious as possible what distinguishes these tokens, when we are using each type, and what we recommend you to configure. The page is now slightly reorganized, and I attempted to cut down on some of the redundant or obvious language a bit, too, so that there are fewer walls of text.

Since I don't think you all can preview what the page looks like without pulling and running the docsite yourself, here's a giant scrolled screenshot of the new page:

screencapture-localhost-5080-batch-changes-how-tos-configuring-credentials-2021-07-27-20_54_44

**Comment by mrnugget** [comment]
> Since I don't think you all can preview what the page looks like without pulling and running the docsite yourself, here's a giant scrolled screenshot of the new page:

This. is. amazing!

**Review by mrnugget — APPROVED** [review]

**Review by LawnGnome — APPROVED** [review]
This is awesome! Great job. :lock_with_ink_pen:
